src/hash.ts

```diff
@@ -3,52 +3,130 @@ const createHash = require('create-hash')
 const ethjsUtil = require('ethjs-util')
 import * as rlp from 'rlp'
 import { toBuffer, setLengthLeft } from './bytes'
+import { assertIsString, assertIsBuffer, assertIsArray } from './helpers'
 
 /**
- * Creates Keccak hash of the input
- * @param a The input data (Buffer|Array|String|Number) If the string is a 0x-prefixed hex value
- * it's interpreted as hexadecimal, otherwise as utf8.
- * @param bits The Keccak width
+ * Creates Keccak hash of a Buffer input
+ * @param a The input data (Buffer)
+ * @param bits (number = 256) The Keccak width
  */
-export const keccak = function(a: any, bits: number = 256): Buffer {
-  if (typeof a === 'string' && !ethjsUtil.isHexString(a)) {
-    a = Buffer.from(a, 'utf8')
-  } else {
-    a = toBuffer(a)
-  }
-
-  if (!bits) bits = 256
-
+export const keccak = function(a: Buffer, bits: number = 256): Buffer {
+  assertIsBuffer(a)
   return createKeccakHash(`keccak${bits}`)
     .update(a)
     .digest()
 }
 
 /**
  * Creates Keccak-256 hash of the input, alias for keccak(a, 256).
- * @param a The input data (Buffer|Array|String|Number)
+ * @param a The input data (Buffer)
  */
-export const keccak256 = function(a: any): Buffer {
+export const keccak256 = function(a: Buffer): Buffer {
   return keccak(a)
 }
 
 /**
- * Creates SHA256 hash of the input.
- * @param a The input data (Buffer|Array|String|Number)
+ * Creates Keccak hash of a String input
+ * @param a The input data (String) If string is 0x-prefixed hex value it's
+ *          interpreted as hexadecimal, otherwise as utf8.
+ * @param bits (number = 256) The Keccak width
+ */
+export const keccakFromString = function(a: string, bits: number = 256) {
+  assertIsString(a)
+  let buf
+  if (!ethjsUtil.isHexString(a)) {
+    buf = Buffer.from(a, 'utf8')
+  } else {
+    buf = toBuffer(a)
+  }
```

Comment:
This leads to ambiguous type interpretation situations, where it is not decidable if a string is a string randomly having a `0x` at the beginning or really being a hex string (let's take e.g. the someone wants to hash "0x is my favorite defi protocol").

Maybe sounds constructed but we had these cases (and problems with it) in the past.

First thought I had was to just take the utf8 case here and leave the hex string case to the buffer version of the method and require users to manually convert.

Second thought, which I would actually prefer since it's a common use case: would suggest to do a separate method `keccakFromHexString` with an `assertHexSrring` check and leave this method for the utf8 case. 

Or do you have a strong opinion here in another direction?

Reply:
I think your approach is right. This is consistent with the recent changes as well. Will update.

Reply:
dbda438531f6e0d861db30a3efb1ea17b43efa8e